### ISSUE TITLE
Error when trying to print from SSH connected site

### ISSUE THREAD

**Issue by MeestorX** [PDConSec/vsc-print#115]
I installed it via SSH as per the instructions.
When clicking the print icon, I get this error:
```
Error while attempting to print: Command failed: xdg-open http://localhost:44007/ /bin/sh: 1: xdg-open: not found
```

VS is on MacOS Big Sur
Client is Ubunutu 20.04, connected to via SSH in VSCode

**Comment by PeterWone** [comment]
It's failing to launch your web browser. `xdg-open` is a _Linux_ command used to launch the default web browser, but going on what you just wrote your workstation runs MacOS and the remote computer runs Ubuntu.

Configure the extension to use an alternate browser and tell it where your browser is. That should work around the problem while I figure out how to test this (I don't own a Mac).

**Comment by MeestorX** [comment]
Setting the alternate browser in "User" and "SSH" to /Applications/Safari.app and I get this error:
```
Error while attempting to print: Command failed: /Applications/Safari.app http://localhost:41151/ /bin/sh: 1: /Applications/Safari.app: not found
```
Looks like it's still trying to find the browser on the remote rather than my local machine...
If I try it while NOT connected to a remote, I get this error:
```
Error while attempting to print: Command failed: /Applications/Safari.app http://localhost:62885/ /bin/sh: /Applications/Safari.app: is a directory
```

BTW, this always worked fine in the past, and I'm using VSCode - Insiders, so perhaps something has broken in the latest update? 

I also don't remember having to "install" this on the remote in the past. Has that always been the case?

**Comment by MeestorX** [comment]
![image](https://user-images.githubusercontent.com/29129658/134427229-a2ad8c24-7b56-4f34-bfda-b4d2b72f888f.png)

**Comment by PeterWone** [comment]
Thanks. And now that I know you're a technical Mac user (versus what I call appliance users) I won't dumb it down and this can go much faster. I'm hamstrung by Apple's obstruction of OSX in VMs, makes me dependent on you for testing and diagnostics. As you can imagine the missus won't sit still for buying a Mac to test software that doesn't earn.

Just a sanity check to make sure we're on the same page, 

This is local

![image](https://user-images.githubusercontent.com/5498936/134428414-d9acdf1a-fe4a-439b-a159-2f362f7201d7.png)

Clicking the green corner button triggers this

![image](https://user-images.githubusercontent.com/5498936/134428605-3acf9987-e97e-411c-8d90-bb60f8648938.png)

and after I connect to the remote host it looks like this

![image](https://user-images.githubusercontent.com/5498936/134428688-3f765a89-ffba-454d-809a-20f9ded8983a.png)

In this connected state, looking at my extensions some of them are offering to put themselves on the remote host.

![image](https://user-images.githubusercontent.com/5498936/134428852-8a773485-2d2d-4165-afd0-e46bf73d26ef.png)

And to do that I just click on them on the label offering.

Is that how it went for you?

**Comment by MeestorX** [comment]
Yes, that's how it works for me as well.

Here's the extensions page:
![image](https://user-images.githubusercontent.com/29129658/134431003-5715562c-0187-4cab-8f59-0d1344b879b5.png)

**Comment by PeterWone** [comment]
Well damn. That looks like it's in order. I'll get Insiders and see whether they've messed me up.

**Comment by MeestorX** [comment]
Sorry, I've not tried that. I know you can install OSX using VMWare or VirtualBox, but I don't run Linux desktop so I've not tried that on Linux.

Here's some google results...
https://www.makeuseof.com/tag/macos-windows-10-virtual-machine/
https://www.soupbowl.io/2020/04/macos-in-virtualbox/

**Comment by PeterWone** [comment]
Since this may take some time to figure out, maybe you should revert to 0.9.8

You'll have to suppress auto-update or it will keep rolling forward. Let me know if you need guidance to do that.

**Comment by PeterWone** [comment]
Just to make things extra weird, when I spun up Insiders, it had the green remoting button at bottom left, then I checked the version which was 1.57 so I let it update, and now there's no remoting button.

**Comment by MeestorX** [comment]
Reverting to 0.9.8 doesn't fix it. TBH, I may have been mistaken when I said that it did work remotely before. I really thought it did, but that also could have been before I updated to Big Sur.

That's weird about insider!

Thanks for your help with this. I'm using VSCode as my "one-stop-shop" for editing, FTP, Terminal, etc. so it would be great to be able to print as well!

**Comment by PeterWone** [comment]
It is a point of pride for me that the Print extension should _just work_. 

With Insider, I looked at extensions, found that none of the remoting extensions were installed and installed for SSH, then connected to one of my Linux hosts, then installed Printing on the remote host as we discussed.

Naturally it works perfectly on the first try without so much as a restart of VS Code. You cannot imagine how much it galls me to say "it works on my machines" because it _really_ pisses me off when people say that to me. We'll get to the bottom of it, hopefully before it produces a bad review.

**Comment by PeterWone** [comment]
Do you have any other computers you can use to test a clean installation?

**Comment by MeestorX** [comment]
Hi Peter;

I apprecaite your passion and attention to "getting it right". No doubt since nobody else is complaining, and it works correctly on your setup, it's something specific to my installation. Your assistance in figuring this out would be greatly appreciated.

Setting up the SSH keys and things on another computer might take some time but it's not impossible. I assume you don't have an idea at the moment what might be causing the issue? As you thought originally, without changing the setup, it seems to want to print at the remote client, not on my dev machine. What setting(s) did you change in your extension to make it show up in your local browser?

**Comment by PeterWone** [comment]
I just built a clean Ubuntu VM and set up vscode, then installed VS Code using the point and click Ubuntu GUI app you get at first login. Then I did OS updates, then a restart, then launched vscode.

Installed Print extension, then Remote SSH. 

Then set up an SSH connection to dockerhost (another Ubuntu VM) and found that Print extension was already present on the remote host probably due to my installing it for connections from my win10 workstation.

The only source code there is the source for the extension itself so I opened that, clicked on one of the files and tried to print.
Up came Firefox (which is the default browser).

![image](https://user-images.githubusercontent.com/5498936/134442565-dcc1991d-3fef-4bdb-ae34-4f011daaf0c8.png)

**Comment by PeterWone** [comment]
I hesitate to say that it's your workstation that's messed up because it's not impossible that this is an OSX thing, but it's hard to prove that what with Apple being a pack of tossers about VMs. <rant>Opening up and supporting developers while providing a coherent software platform is the reason Microsoft dominates the desktop. Apple's position is not supported by reality.</rant>

At any rate I see two paths open to you. 

1. Remove VS Code (not Insiders) 
2. Remove the `~/.vscode` folder if it's still there. This will remove all the extensions that were installed on VSCode but not Insiders.
3. Install VS Code
4. Install Print extension
5. Install Remote SSH (or the whole Remote Development pack if you like)
6. Click the remoting button then connect to your remote host. If you already set up key login it will go straight through, if that doesn't work it will prompt for your password
7. Inspect your extensions for a look at what's installed on the remote host, Print should already be there from earlier efforts.
8. Open something on the remote host and try to print.
9. Rejoice or cry depending on outcome.

**Comment by PeterWone** [comment]
Incidentally Chromium based browsers seem to work the best for printing. In reviews people mention success using Brave on Macs.

**Comment by PeterWone** [comment]
Hey you said reverting to 0.9.8 _didn't_ fix it. 

Try this

1. Start VS Code Insiders
2. Uninstall Print extension
3. Stop VS Code Insiders (all instances)
4. In `~/.vscode-insiders` you should find folders for the extensions you have installed. Look for pdconsec.vscode-print-0.9.8 and any other versions. Delete all the folders `~/.vscode-insiders/pdconsec.vscode-print-*`
5. Start VS Code Insiders and look at your extensions.
6. Print extension should _not_ be present. Install it from the net.
7. Test it locally.
8. Test it remote.

**Comment by PeterWone** [comment]
If you ever do come back to this, don't hesitate to re-open this issue.

**Comment by PeterWone** [comment]
@gerritgriebel I've re-opened this because a second person experiencing the same thing in similar conditions indicates a real problem.

Happily we are in a much stronger position to troubleshoot now. 

1. In the Printing settings, change your log level to `debug`
2. Restart VS Code and reproduce the failure.
3. Change your log level back to warn or error so it doesn't grow too fast
4. Open the print log and copy the tail to a post here so I can have a look.
![image](https://user-images.githubusercontent.com/5498936/192792612-b4291434-c650-4b4b-af9a-5bc0c7b1ffe9.png)

I'm wondering whether your remote host is Linux and that your log will show that VS Code is reporting the platform as Linux because the extension is running in the proxy on the remote host to which you are SSHing. That might result in the shell command being the wrong one for the platform on which the browser must launch.

**Comment by gerritgriebel** [comment]
It's just the same message as in popup: 

<img width="382" alt="Bildschirmfoto 2022-09-29 um 08 01 14" src="https://user-images.githubusercontent.com/3919675/192951317-dab1dced-159a-4681-adf9-1d74c06e15ba.png">

```
2022-09-29 07:57:05 debug: Print activated
2022-09-29 07:57:05 info: Began listening on 127.0.0.1:45215
2022-09-29 07:57:49 info: Garbage collection stopped by deactivate
2022-09-29 07:58:11 debug: Print activated
2022-09-29 07:58:11 info: Began listening on 127.0.0.1:39625
2022-09-29 07:58:14 debug: Print command was invoked
2022-09-29 07:58:14 debug: Creating a print session object for file:///home/egdev/webapp/test.txt
2022-09-29 07:58:14 debug: Using the buffer of the active editor
2022-09-29 07:58:14 debug: Source code line numbers will be printed
2022-09-29 07:58:14 debug: Source code colour scheme is "Atelier Dune"
2022-09-29 07:58:14 debug: Platform detected as "linux" so launch browser command is "/usr/bin/open"
2022-09-29 07:58:14 error: Error while attempting to print: Command failed: /usr/bin/open http://localhost:39625/15692cd2-a3e3-4f36-8a7b-47fea2793d1f/
/bin/sh: 1: /usr/bin/open: not found
```
Local platform ist macOS remote platform is Debian Linux. Command to open page on macOS side should be /usr/bin/open which opens URL on first parameter position with standard browser

**Comment by PeterWone** [comment]
I finally had a chance to use OSX to run the extension in conjunction with an SSH workspace running on Ubuntu, which is a Debian derivative. Annoyingly, "It works on my machine"(TM)

```
2022-10-02 13:21:50 debug: Print activated
2022-10-02 13:21:50 info: Began listening on 127.0.0.1:46851
2022-10-02 13:22:27 debug: Print command was invoked
2022-10-02 13:22:27 debug: Creating a print session object for file:///home/peter/.vscode-server/extensions/pdconsec.vscode-print-0.10.4/vscode-print.log
2022-10-02 13:22:27 debug: Using the buffer of the active editor
2022-10-02 13:22:27 debug: Source code line numbers will be printed
2022-10-02 13:22:27 debug: Source code colour scheme is "Atelier Dune"
2022-10-02 13:22:27 debug: Platform detected as "linux" so launch browser command is "xdg-open"
2022-10-02 13:22:29 debug: Responding to base document request for session 0d1d311f-e7c1-43c1-9495-36a090fe4d40
2022-10-02 13:22:29 debug: Printing /home/peter/.vscode-server/extensions/pdconsec.vscode-print-0.10.4/vscode-print.log
2022-10-02 13:22:30 debug: Responding to vsc-print.resource request for default.css in session 0d1d311f-e7c1-43c1-9495-36a090fe4d40
2022-10-02 13:22:31 debug: Responding to vsc-print.resource request for line-numbers.css in session 0d1d311f-e7c1-43c1-9495-36a090fe4d40
2022-10-02 13:22:31 debug: Responding to vsc-print.resource request for colour-scheme.css in session 0d1d311f-e7c1-43c1-9495-36a090fe4d40
2022-10-02 13:22:31 debug: Loading colour scheme from atelier-dune-light
2022-10-02 13:22:31 debug: Responding to vsc-print.resource request for settings.css in session 0d1d311f-e7c1-43c1-9495-36a090fe4d40
2022-10-02 13:22:31 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:22:31 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:22:31 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:22:31 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:22:31 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:22:59 debug: Responding to "completed" request for session 0d1d311f-e7c1-43c1-9495-36a090fe4d40
2022-10-02 13:23:03 debug: Print command was invoked
2022-10-02 13:23:03 debug: Creating a print session object for file:///home/peter/.vscode-server/extensions/pdconsec.vscode-print-0.10.4/vscode-print.log
2022-10-02 13:23:03 debug: Using the buffer of the active editor
2022-10-02 13:23:03 debug: Source code line numbers will be printed
2022-10-02 13:23:03 debug: Source code colour scheme is "Atelier Dune"
2022-10-02 13:23:03 debug: Platform detected as "linux" so launch browser command is "xdg-open"
2022-10-02 13:23:03 debug: Responding to base document request for session 7e47348c-2e30-42ba-9ec2-6cfa973fac3c
2022-10-02 13:23:03 debug: Printing /home/peter/.vscode-server/extensions/pdconsec.vscode-print-0.10.4/vscode-print.log
2022-10-02 13:23:05 debug: Responding to vsc-print.resource request for default.css in session 7e47348c-2e30-42ba-9ec2-6cfa973fac3c
2022-10-02 13:23:05 debug: Responding to vsc-print.resource request for line-numbers.css in session 7e47348c-2e30-42ba-9ec2-6cfa973fac3c
2022-10-02 13:23:05 debug: Responding to vsc-print.resource request for colour-scheme.css in session 7e47348c-2e30-42ba-9ec2-6cfa973fac3c
2022-10-02 13:23:05 debug: Loading colour scheme from atelier-dune-light
2022-10-02 13:23:05 debug: Responding to vsc-print.resource request for settings.css in session 7e47348c-2e30-42ba-9ec2-6cfa973fac3c
2022-10-02 13:23:06 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:23:06 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:23:06 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:23:06 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:23:06 warn: Dropping connection of /favicon.ico does not correspond to a print session
2022-10-02 13:24:08 debug: Responding to "completed" request for session 7e47348c-2e30-42ba-9ec2-6cfa973fac3c
```

Now, it's logging the OS as linux, as yours does, with a default browser shell command of `xdg-open` which is right for Linux and wrong for OSX (I checked, OSX does not understand `xdg-open`). Since it works on my machine, with the browser correctly launching on the OSX notebook, it seems like code producing the log message is running in the proxy on the SSH host but the browser is being launched by code running on the workstation (this is what I intended).

So I will fix up the logging of the platform and launch command so that it runs in the same context as the browser launch - even if I have to inline it with browser launch. Then if you update and re-run the test the log will tell us what is going on on **your** workstations.

**Comment by gerritgriebel** [comment]
ok, thanks! Now it works and I got it: /usr/bin/xdg-open is actually needed on server (linux) side to make it work, so installing it helped:
```
# apt-get install xdg-utils
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following additional packages will be installed:
  cpp cpp-8 libdrm-amdgpu1 libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libfile-basedir-perl libfile-desktopentry-perl libfile-mimeinfo-perl libfontenc1 libgl1 libgl1-mesa-dri libglapi-mesa libglvnd0
  libglx-mesa0 libglx0 libice6 libipc-system-simple-perl libisl19 libllvm7 libmpc3 libnet-dbus-perl libpciaccess0 libsensors-config libsensors5 libsm6 libtie-ixhash-perl libx11-protocol-perl libx11-xcb1
  libxaw7 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0 libxcb-shape0 libxcb-sync1 libxft2 libxml-twig-perl libxml-xpathengine-perl libxmu6 libxpm4 libxshmfence1 libxt6 libxtst6 libxv1
  libxxf86dga1 libxxf86vm1 x11-common x11-utils x11-xserver-utils
```
But: why? I had to install a bunch of dependencies like x11 libs. IMHO, it should not be necessary for this task of producing a HTML page that is rendered in client side browser. Is there a smarter way than using default /usr/bin/xdg-open?

**Comment by PeterWone** [comment]
You've indirectly explained why it does work on my system; looks like execution is delegated back to the workstation via x11. The installation of Ubuntu I've been using had all that stuff already set up.

That makes sense. I already knew the markdown pipeline runs on the remote host proxy.

**Comment by PeterWone** [comment]
I'm going to close this issue because we got to the bottom of the problem, and open another issue "eliminate dependency on xdg-open"
